File: bindings/python/src/datalevin/connection.py
"""High-level Python wrapper for Datalevin Datalog connections."""

from __future__ import annotations

from concurrent.futures import Future
from threading import Thread

import jpype

from ._convert import to_edn_form, to_java, to_python, to_query_input
from ._interop import _BINDINGS
from ._java import call_java, classes
from ._resource import ResourceWrapper


def _edn_form(value):
    if isinstance(value, str):
        return _BINDINGS.read_edn(value)
    return to_edn_form(value)


def _slice_rows(rows, limit=None, offset=0):
    start = max(offset or 0, 0)
    if limit is None:
        return rows[start:] if start else rows
    return rows[start : start + max(limit, 0)]


def _fetch_limit(limit, offset=0):
    if limit is None:
        return None
    return max(limit, 0) + max(offset or 0, 0)


def _python_future(java_future):
    future = Future()

    def complete():
        try:
            future.set_result(to_python(call_java(java_future.get)))
        except BaseException as exc:
            future.set_exception(exc)

    Thread(target=complete, daemon=True).start()
    return future


class _PythonConsumer:
    def __init__(self, fn) -> None:
        self._fn = fn

    def accept(self, value):
        self._fn(to_python(value))


def _consumer_proxy(fn):
    if not callable(fn):
        raise TypeError("callback must be callable.")
    return jpype.JProxy(classes().consumer_type, inst=_PythonConsumer(fn))


class Connection(ResourceWrapper):
    """Thin Python wrapper over a raw Datalevin connection handle."""

    def __init__(self, handle, *, owned: bool = True) -> None:
        close_fn = _BINDINGS.close_connection if owned else lambda _handle: None
        super().__init__(handle, close_fn, _BINDINGS.connection_closed, "connection")
        self._listeners = {}

    def close(self) -> None:
        self._listeners.clear()
        super().close()

    @staticmethod
    def _query_input(value):
        if isinstance(value, Connection):
            return _BINDINGS.connection_db(value.raw_handle())
        return to_query_input(value)

    def schema(self):
        return to_python(_BINDINGS.core_invoke("schema", [self.raw_handle()]))

    def opts(self):
        return to_python(_BINDINGS.core_invoke("opts", [self.raw_handle()]))

    def update_schema(self, schema_update, del_attrs=None, rename_map=None):
        args = [self.raw_handle(), _BINDINGS.schema(schema_update) if schema_update is not None else None]
        if rename_map is not None:
            args.append(_BINDINGS.delete_attrs(del_attrs))
            args.append(_BINDINGS.rename_map(rename_map))
        elif del_attrs is not None:
            args.append(_BINDINGS.delete_attrs(del_attrs))
        return to_python(_BINDINGS.core_invoke("update-schema", args))

    def clear(self) -> None:
        _BINDINGS.core_invoke("clear", [self.raw_handle()])

    def with_transaction(self, fn):
        return to_python(_BINDINGS.connection_with_transaction(self.raw_handle(), fn))

    def entid(self, eid):
        db = _BINDINGS.connection_db(self.raw_handle())
        return to_python(_BINDINGS.core_invoke("entid", [db, _BINDINGS.lookup_ref(eid)]))

    def entity(self, eid):
        from .entity import Entity

        entity = _BINDINGS.connection_entity(self.raw_handle(), eid)
        if entity is None:
            return None
        return Entity(entity)

    def entity_map(self, eid):
        entity = _BINDINGS.connection_entity(self.raw_handle(), eid)
        if entity is None:
            return None
        return to_python(_BINDINGS.entity_touch(entity))

    def pull(self, selector, eid):
        db = _BINDINGS.connection_db(self.raw_handle())
        return to_python(
            _BINDINGS.core_invoke("pull", [db, _edn_form(selector), _BINDINGS.lookup_ref(eid)])
        )

    def pull_many(self, selector, eids):
        db = _BINDINGS.connection_db(self.raw_handle())
        refs = [_BINDINGS.lookup_ref(eid) for eid in eids]
        return to_python(_BINDINGS.core_invoke("pull-many", [db, _edn_form(selector), refs]))

    def query(self, query, *inputs):
        db = _BINDINGS.connection_db(self.raw_handle())
        args = [_edn_form(query), db, *(self._query_input(value) for value in inputs)]
        return to_python(_BINDINGS.core_invoke("q", args))

    def explain(self, query, *inputs, opts_edn=None):
        db = _BINDINGS.connection_db(self.raw_handle())
        opts = None if opts_edn is None else _edn_form(opts_edn)
        args = [opts, _edn_form(query), db, *(self._query_input(value) for value in inputs)]
        return to_python(_BINDINGS.core_invoke("explain", args))

    def transact(self, tx_data, tx_meta=None):
        args = [self.raw_handle(), _BINDINGS.tx_data(tx_data)]
        if tx_meta is not None:
            args.append(to_java(tx_meta))
        return to_python(_BINDINGS.core_invoke("transact!", args))

    def transact_async(self, tx_data, tx_meta=None):
        """Start an async transaction and return a concurrent.futures.Future."""

        return _python_future(
            _BINDINGS.connection_transact_async(self.raw_handle(), tx_data, tx_meta)
        )

    def listen(self, callback, key=None):
        proxy = _consumer_proxy(callback)
        if key is None:
            registered_key = to_python(_BINDINGS.connection_listen(self.raw_handle(), proxy))
        else:
            registered_key = to_python(_BINDINGS.connection_listen(self.raw_handle(), key, proxy))
        self._listeners[registered_key] = proxy
        return registered_key

    def unlisten(self, key) -> None:
        _BINDINGS.connection_unlisten(self.raw_handle(), key)
        self._listeners.pop(key, None)

    def fill_db(self, datoms):
        _BINDINGS.fill_db(self.raw_handle(), datoms)
        return self

    def datalog_kv(self):
        from .kv import KV

        return KV(_BINDINGS.connection_datalog_kv(self.raw_handle()), owned=False)

    def datoms(self, index, c1=None, c2=None, c3=None, limit=None, offset=0):
        rows = to_python(
            _BINDINGS.connection_datoms(
                self.raw_handle(),
                index,
                c1,
                c2,
                c3,
                _fetch_limit(limit, offset),
            )
        )
        return _slice_rows(rows, limit, offset)

    def search_datoms(self, e=None, attr=None, value=None, limit=None, offset=0):
        rows = to_python(_BINDINGS.connection_search_datoms(self.raw_handle(), e, attr, value))
        return _slice_rows(rows, limit, offset)

    def count_datoms(self, e=None, attr=None, value=None):
        return to_python(_BINDINGS.connection_count_datoms(self.raw_handle(), e, attr, value))

    def seek_datoms(self, index, c1=None, c2=None, c3=None, limit=None, offset=0):
        rows = to_python(
            _BINDINGS.connection_seek_datoms(
                self.raw_handle(),
                index,
                c1,
                c2,
                c3,
                _fetch_limit(limit, offset),
            )
        )
        return _slice_rows(rows, limit, offset)

    def rseek_datoms(self, index, c1=None, c2=None, c3=None, limit=None, offset=0):
        rows = to_python(
            _BINDINGS.connection_rseek_datoms(
                self.raw_handle(),
                index,
                c1,
                c2,
                c3,
                _fetch_limit(limit, offset),
            )
        )
        return _slice_rows(rows, limit, offset)

    def index_range(self, attr, start, end, limit=None, offset=0):
        rows = to_python(_BINDINGS.connection_index_range(self.raw_handle(), attr, start, end))
        return _slice_rows(rows, limit, offset)

    def fulltext_datoms(self, query, opts=None, limit=None, offset=0):
        rows = to_python(_BINDINGS.connection_fulltext_datoms(self.raw_handle(), query, opts))
        return _slice_rows(rows, limit, offset)

    def copy(self, dest, compact=None) -> None:
        _BINDINGS.connection_copy(self.raw_handle(), dest, compact)

    def tx_log_watermarks(self):
        return to_python(_BINDINGS.connection_tx_log_watermarks(self.raw_handle()))

    def open_tx_log(self, from_lsn, upto_lsn=None, limit=None):
        rows = to_python(_BINDINGS.connection_open_tx_log(self.raw_handle(), from_lsn, upto_lsn))
        if limit is None:
            return rows
        return rows[: max(limit, 0)]

    def create_snapshot(self):
        return to_python(_BINDINGS.connection_create_snapshot(self.raw_handle()))

    def list_snapshots(self):
        return to_python(_BINDINGS.connection_list_snapshots(self.raw_handle()))

    def gc_tx_log_segments(self, retain_floor_lsn=None):
        return to_python(_BINDINGS.connection_gc_tx_log_segments(self.raw_handle(), retain_floor_lsn))

    def re_index(self, opts=None, *, schema=None):
        self._handle = _BINDINGS.connection_re_index(self.raw_handle(), schema, opts)
        return self

File: bindings/javascript/src/connection.js
import { toEdnForm, toJava, toJs, toQueryInput } from "./convert.js";
import { txReportToJs } from "./database.js";
import { _BINDINGS } from "./interop.js";
import { callJavaMethod, javaBridgeModule } from "./jvm.js";
import { ResourceWrapper } from "./resource.js";
import { toJsResult } from "./result.js";

async function pullSelector(value) {
  if (typeof value === "string") {
    return value;
  }
  return toEdnForm(value);
}

async function queryForm(query) {
  if (typeof query !== "string") {
    return toEdnForm(query);
  }
  return query;
}

function hasValue(value) {
  return value !== null && value !== undefined;
}

function sliceRows(rows, { limit = null, offset = 0 } = {}) {
  const start = Math.max(offset ?? 0, 0);
  if (!hasValue(limit)) {
    return start === 0 ? rows : rows.slice(start);
  }
  return rows.slice(start, start + Math.max(limit, 0));
}

function fetchLimit(limit, offset = 0) {
  if (!hasValue(limit)) {
    return null;
  }
  return Math.max(limit, 0) + Math.max(offset ?? 0, 0);
}

function fulltextOpts(opts = null, { limit = null, offset = 0, pagingCachePages = null } = {}) {
  const mapOpts = opts instanceof Map;
  const merged = mapOpts ? new Map(opts) : (hasValue(opts) ? { ...opts } : {});
  const put = (key, value) => {
    if (mapOpts) {
      merged.set(key, value);
    } else {
      merged[key] = value;
    }
  };
  if (hasValue(limit)) {
    put(":limit", Math.max(limit, 0));
  }
  if (offset) {
    put(":offset", Math.max(offset, 0));
  }
  if (hasValue(pagingCachePages)) {
    put(":paging-cache-pages", pagingCachePages);
  }
  return (mapOpts ? merged.size === 0 : Object.keys(merged).length === 0) ? null : merged;
}

let listenerProxyEventLoopDepth = 0;
let listenerProxyEventLoopPrevious = false;

async function retainListenerProxyEventLoop() {
  const bridge = await javaBridgeModule();
  if (listenerProxyEventLoopDepth === 0) {
    listenerProxyEventLoopPrevious = bridge.config.runEventLoopWhenInterfaceProxyIsActive;
    bridge.config.runEventLoopWhenInterfaceProxyIsActive = true;
  }
  listenerProxyEventLoopDepth += 1;
  return () => {
    listenerProxyEventLoopDepth -= 1;
    if (listenerProxyEventLoopDepth === 0) {
      bridge.config.runEventLoopWhenInterfaceProxyIsActive = listenerProxyEventLoopPrevious;
    }
  };
}

async function createConsumerProxy(fn) {
  if (typeof fn !== "function") {
    throw new TypeError("callback must be a function.");
  }
  const { newProxy } = await javaBridgeModule();
  return newProxy("java.util.function.Consumer", {
    accept: async (value) => {
      await fn(await toJs(value));
    }
  });
}

export class Connection extends ResourceWrapper {
  constructor(handle, { owned = true } = {}) {
    super(
      handle,
      owned ? (rawHandle) => _BINDINGS.closeConnection(rawHandle) : async () => {},
      (rawHandle) => _BINDINGS.connectionClosed(rawHandle),
      "connection"
    );
    this._listenerProxies = new Map();
  }

  async close() {
    this._resetListenerProxies();
    await super.close();
  }

  _resetListenerProxies() {
    for (const { proxy, release } of this._listenerProxies.values()) {
      proxy.reset?.();
      release?.();
    }
    this._listenerProxies.clear();
  }

  async schema() {
    return toJsResult(await _BINDINGS.coreInvoke("schema", [this.rawHandle()]));
  }

  async opts() {
    return toJsResult(await _BINDINGS.coreInvoke("opts", [this.rawHandle()]));
  }

  async datalogIndexCacheLimit(limit = null) {
    if (hasValue(limit)) {
      return toJsResult(
        await callJavaMethod(this.rawHandle(), "datalogIndexCacheLimit", await toJava(limit))
      );
    }
    return toJsResult(await callJavaMethod(this.rawHandle(), "datalogIndexCacheLimit"));
  }

  async maxEid() {
    return toJsResult(await callJavaMethod(this.rawHandle(), "maxEid"));
  }

  async cardinality(attr) {
    return toJsResult(await callJavaMethod(this.rawHandle(), "cardinality", await toJava(attr)));
  }

  async analyze(attr = null) {
    if (hasValue(attr)) {
      return toJsResult(await callJavaMethod(this.rawHandle(), "analyze", await toJava(attr)));
    }
    return toJsResult(await callJavaMethod(this.rawHandle(), "analyze"));
  }

  async updateSchema(schemaUpdate, { delAttrs = null, renameMap = null } = {}) {
    const args = [
      this.rawHandle(),
      schemaUpdate === null || schemaUpdate === undefined ? null : await _BINDINGS.schema(schemaUpdate)
    ];

    if (renameMap !== null && renameMap !== undefined) {
      args.push(await _BINDINGS.deleteAttrs(delAttrs));
      args.push(await _BINDINGS.renameMap(renameMap));
    } else if (delAttrs !== null && delAttrs !== undefined) {
      args.push(await _BINDINGS.deleteAttrs(delAttrs));
    }

    return toJsResult(await _BINDINGS.coreInvoke("update-schema", args));
  }

  async clear() {
    await _BINDINGS.coreInvoke("clear", [this.rawHandle()]);
  }

  async entid(eid) {
    return toJsResult(await callJavaMethod(this.rawHandle(), "entid", await toJava(eid)));
  }

  async entity(eid) {
    const { Entity } = await import("./entity.js");
    const entity = await _BINDINGS.connectionEntity(this.rawHandle(), eid);
    return entity === null || entity === undefined ? null : new Entity(entity);
  }

  async entityMap(eid) {
    return toJsResult(await callJavaMethod(this.rawHandle(), "entityMap", await toJava(eid)));
  }

  async pull(selector, eid) {
    return toJsResult(
      await callJavaMethod(this.rawHandle(), "pull", await pullSelector(selector), await toJava(eid)),
      { bridge: true }
    );
  }

  async pullMany(selector, eids) {
    return toJsResult(
      await callJavaMethod(this.rawHandle(), "pullMany", await pullSelector(selector), await toJava(eids)),
      { bridge: true }
    );
  }

  async query(query, ...inputs) {
    const normalizedInputs = [];
    for (const input of inputs) {
      normalizedInputs.push(await toQueryInput(input));
    }
    if (typeof query === "string") {
      return toJsResult(
        await callJavaMethod(this.rawHandle(), "query", query, await toJava(normalizedInputs)),
        { bridge: true }
      );
    }
    return toJsResult(
      await callJavaMethod(this.rawHandle(), "queryForm", await queryForm(query), await toJava(normalizedInputs)),
      { bridge: true }
    );
  }

  async explain(query, { inputs = [], optsEdn = null } = {}) {
    const normalizedInputs = [];
    for (const input of inputs) {
      normalizedInputs.push(await toQueryInput(input));
    }
    if (optsEdn !== null && optsEdn !== undefined) {
      if (typeof query === "string") {
        return toJsResult(
          await callJavaMethod(
            this.rawHandle(),
            "explain",
            optsEdn,
            query,
            await toJava(normalizedInputs)
          ),
          { bridge: true }
        );
      }
      return toJsResult(
        await callJavaMethod(
          this.rawHandle(),
          "explainForm",
          optsEdn,
          await queryForm(query),
          await toJava(normalizedInputs)
        ),
        { bridge: true }
      );
    }

    if (typeof query === "string") {
      return toJsResult(
        await callJavaMethod(this.rawHandle(), "explain", query, await toJava(normalizedInputs)),
        { bridge: true }
      );
    }
    return toJsResult(
      await callJavaMethod(this.rawHandle(), "explainForm", await queryForm(query), await toJava(normalizedInputs)),
      { bridge: true }
    );
  }

  async transact(txData, txMeta = null) {
    const args = [this.rawHandle(), await _BINDINGS.txData(txData)];
    if (txMeta !== null && txMeta !== undefined) {
      args.push(await toJava(txMeta));
    }
    return toJsResult(await _BINDINGS.coreInvoke("transact!", args));
  }

  async transactAsync(txData, txMeta = null) {
    const future = await _BINDINGS.connectionTransactAsync(this.rawHandle(), txData, txMeta);
    return toJsResult(await callJavaMethod(future, "get"), { bridge: true });
  }

  async txDataToSimulatedReport(txData) {
    return txReportToJs(await _BINDINGS.connectionTxDataToSimulatedReport(this.rawHandle(), txData));
  }

  async listen(callback, key = null) {
    const proxy = await createConsumerProxy(callback);
    let registeredKey;
    try {
      registeredKey = hasValue(key)
        ? await toJsResult(await _BINDINGS.connectionListen(this.rawHandle(), key, proxy), { bridge: true })
        : await toJsResult(await _BINDINGS.connectionListen(this.rawHandle(), proxy), { bridge: true });
    } catch (error) {
      proxy.reset?.();
      throw error;
    }
    const previous = this._listenerProxies.get(registeredKey);
    previous?.proxy.reset?.();
    previous?.release?.();
    this._listenerProxies.set(registeredKey, {
      proxy,
      release: await retainListenerProxyEventLoop()
    });
    return registeredKey;
  }

  async unlisten(key) {
    await _BINDINGS.connectionUnlisten(this.rawHandle(), key);
    const existing = this._listenerProxies.get(key);
    existing?.proxy.reset?.();
    existing?.release?.();
    this._listenerProxies.delete(key);
  }

  async fillDb(datoms) {
    await callJavaMethod(this.rawHandle(), "fillDb", await toJava(datoms));
    return this;
  }

  async datalogKv() {
    const { KV } = await import("./kv.js");
    return new KV(await callJavaMethod(this.rawHandle(), "datalogKV"), { owned: false });
  }

  async reIndex(opts = null, { schema = null } = {}) {
    this._handle = await _BINDINGS.connectionReIndex(this.rawHandle(), schema, opts);
    return this;
  }

  async datoms(index, { c1 = null, c2 = null, c3 = null, limit = null, offset = 0 } = {}) {
    const capped = fetchLimit(limit, offset);
    const args = [this.rawHandle(), "datoms", await toJava(index), await toJava(c1), await toJava(c2), await toJava(c3)];
    if (hasValue(capped)) {
      args.push(await toJava(capped));
    }
    const rows = await toJsResult(
      await callJavaMethod(...args),
      { bridge: true }
    );
    return sliceRows(rows, { limit, offset });
  }

  async searchDatoms({ e = null, attr = null, value = null, limit = null, offset = 0 } = {}) {
    const rows = await toJsResult(
      await callJavaMethod(this.rawHandle(), "searchDatoms", await toJava(e), await toJava(attr), await toJava(value)),
      { bridge: true }
    );
    return sliceRows(rows, { limit, offset });
  }

  async countDatoms({ e = null, attr = null, value = null } = {}) {
    return toJsResult(
      await callJavaMethod(this.rawHandle(), "countDatoms", await toJava(e), await toJava(attr), await toJava(value))
    );
  }

  async seekDatoms(index, { c1 = null, c2 = null, c3 = null, limit = null, offset = 0 } = {}) {
    const capped = fetchLimit(limit, offset);
    const args = [
      this.rawHandle(),
      "seekDatoms",
      await toJava(index),
      await toJava(c1),
      await toJava(c2),
      await toJava(c3)
    ];
    if (hasValue(capped)) {
      args.push(await toJava(capped));
    }
    const rows = await toJsResult(
      await callJavaMethod(...args),
      { bridge: true }
    );
    return sliceRows(rows, { limit, offset });
  }

  async rseekDatoms(index, { c1 = null, c2 = null, c3 = null, limit = null, offset = 0 } = {}) {
    const capped = fetchLimit(limit, offset);
    const args = [
      this.rawHandle(),
      "rseekDatoms",
      await toJava(index),
      await toJava(c1),
      await toJava(c2),
      await toJava(c3)
    ];
    if (hasValue(capped)) {
      args.push(await toJava(capped));
    }
    const rows = await toJsResult(
      await callJavaMethod(...args),
      { bridge: true }
    );
    return sliceRows(rows, { limit, offset });
  }

  async indexRange(attr, start, end, { limit = null, offset = 0 } = {}) {
    const rows = await toJsResult(
      await callJavaMethod(this.rawHandle(), "indexRange", await toJava(attr), await toJava(start), await toJava(end)),
      { bridge: true }
    );
    return sliceRows(rows, { limit, offset });
  }

  async fulltextDatoms(query, { opts = null, limit = null, offset = 0, pagingCachePages = null } = {}) {
    const searchOpts = fulltextOpts(opts, { limit, offset, pagingCachePages });
    const result = hasValue(searchOpts)
      ? await callJavaMethod(this.rawHandle(), "fulltextDatoms", query, await toJava(searchOpts))
      : await callJavaMethod(this.rawHandle(), "fulltextDatoms", query);
    return toJsResult(
      result,
      { bridge: true }
    );
  }

  async copy(dest, { compact = null } = {}) {
    if (hasValue(compact)) {
      await callJavaMethod(this.rawHandle(), "copy", dest, Boolean(compact));
      return;
    }
    await callJavaMethod(this.rawHandle(), "copy", dest);
  }

  async txLogWatermarks() {
    return toJsResult(await callJavaMethod(this.rawHandle(), "txLogWatermarks"), { bridge: true });
  }

  async openTxLog(fromLsn, { uptoLsn = null, limit = null } = {}) {
    const result = hasValue(uptoLsn)
      ? await callJavaMethod(this.rawHandle(), "openTxLog", await toJava(fromLsn), await toJava(uptoLsn))
      : await callJavaMethod(this.rawHandle(), "openTxLog", await toJava(fromLsn));
    const rows = await toJsResult(result, { bridge: true });
    if (limit === null || limit === undefined) {
      return rows;
    }
    return rows.slice(0, Math.max(limit, 0));
  }

  async createSnapshot() {
    return toJsResult(await callJavaMethod(this.rawHandle(), "createSnapshot"), { bridge: true });
  }

  async listSnapshots() {
    return toJsResult(await callJavaMethod(this.rawHandle(), "listSnapshots"), { bridge: true });
  }

  async gcTxLogSegments({ retainFloorLsn = null } = {}) {
    if (hasValue(retainFloorLsn)) {
      return toJsResult(
        await callJavaMethod(this.rawHandle(), "gcTxLogSegments", await toJava(retainFloorLsn)),
        { bridge: true }
      );
    }
    return toJsResult(await callJavaMethod(this.rawHandle(), "gcTxLogSegments"), { bridge: true });
  }
}
